Error: SQLAlchemy SSL connection has been closed unexpectedly

We've now seen two 5xx errors today from the API around the RDS proxy SSL connection closing unexpectedly:
```
sqlalchemy.exc.OperationalError: (psycopg2.OperationalError) SSL connection has been closed unexpectedly
```
This may be a similar problem to what [Notify was facing when they switched over to the RDS Proxy](https://gcdigital.slack.com/archives/CV38DBNVA/p1631050570033600), which was fixed by increasing the RDS proxy idle connection timeout.

That being said, the ListManager RDS proxy is already using an [idle_client_timeout of 1800](https://github.com/cds-snc/terraform-modules/blob/v0.0.35/rds/main.tf#L77).





Full trace is:
```
  | 2021-09-16T13:51:03.428-04:00 | [ERROR] 2021-09-16T17:51:03.315Z 273164e9-a0aa-4bdc-97e4-ee2a5fe1d4b3 Exception in 'http' protocol.
  | 2021-09-16T13:51:03.428-04:00 | Traceback (most recent call last):
  | 2021-09-16T13:51:03.428-04:00 | File "/pymodules/mangum/protocols/http.py", line 81, in run
  | 2021-09-16T13:51:03.428-04:00 | await app(self.request.scope, self.receive, self.send)
  | 2021-09-16T13:51:03.428-04:00 | File "/pymodules/fastapi/applications.py", line 199, in __call__
  | 2021-09-16T13:51:03.428-04:00 | await super().__call__(scope, receive, send)
  | 2021-09-16T13:51:03.428-04:00 | File "/pymodules/starlette/applications.py", line 112, in __call__
  | 2021-09-16T13:51:03.428-04:00 | await self.middleware_stack(scope, receive, send)
  | 2021-09-16T13:51:03.428-04:00 | File "/pymodules/starlette/middleware/errors.py", line 181, in __call__
  | 2021-09-16T13:51:03.428-04:00 | raise exc from None
  | 2021-09-16T13:51:03.428-04:00 | File "/pymodules/starlette/middleware/errors.py", line 159, in __call__
  | 2021-09-16T13:51:03.428-04:00 | await self.app(scope, receive, _send)
  | 2021-09-16T13:51:03.428-04:00 | File "/pymodules/starlette/exceptions.py", line 82, in __call__
  | 2021-09-16T13:51:03.428-04:00 | raise exc from None
  | 2021-09-16T13:51:03.428-04:00 | File "/pymodules/starlette/exceptions.py", line 71, in __call__
  | 2021-09-16T13:51:03.428-04:00 | await self.app(scope, receive, sender)
  | 2021-09-16T13:51:03.428-04:00 | File "/pymodules/starlette/routing.py", line 580, in __call__
  | 2021-09-16T13:51:03.428-04:00 | await route.handle(scope, receive, send)
  | 2021-09-16T13:51:03.428-04:00 | File "/pymodules/starlette/routing.py", line 241, in handle
  | 2021-09-16T13:51:03.428-04:00 | await self.app(scope, receive, send)
  | 2021-09-16T13:51:03.428-04:00 | File "/pymodules/starlette/routing.py", line 52, in app
  | 2021-09-16T13:51:03.428-04:00 | response = await func(request)
  | 2021-09-16T13:51:03.428-04:00 | File "/pymodules/fastapi/routing.py", line 219, in app
  | 2021-09-16T13:51:03.428-04:00 | raw_response = await run_endpoint_function(
  | 2021-09-16T13:51:03.428-04:00 | File "/pymodules/fastapi/routing.py", line 154, in run_endpoint_function
  | 2021-09-16T13:51:03.428-04:00 | return await run_in_threadpool(dependant.call, **values)
  | 2021-09-16T13:51:03.428-04:00 | File "/pymodules/starlette/concurrency.py", line 40, in run_in_threadpool
  | 2021-09-16T13:51:03.428-04:00 | return await loop.run_in_executor(None, func, *args)
  | 2021-09-16T13:51:03.428-04:00 | File "/usr/local/lib/python3.9/concurrent/futures/thread.py", line 52, in run
  | 2021-09-16T13:51:03.428-04:00 | result = self.fn(*self.args, **self.kwargs)
  | 2021-09-16T13:51:03.428-04:00 | File "/function/api_gateway/api.py", line 145, in get_list_counts
  | 2021-09-16T13:51:03.428-04:00 | lists = session.query(List.id).filter(List.service_id == service_id).all()
  | 2021-09-16T13:51:03.428-04:00 | File "/pymodules/sqlalchemy/orm/query.py", line 2712, in all
  | 2021-09-16T13:51:03.428-04:00 | return self._iter().all()
  | 2021-09-16T13:51:03.428-04:00 | File "/pymodules/sqlalchemy/orm/query.py", line 2847, in _iter
  | 2021-09-16T13:51:03.428-04:00 | result = self.session.execute(
  | 2021-09-16T13:51:03.428-04:00 | File "/pymodules/sqlalchemy/orm/session.py", line 1689, in execute
  | 2021-09-16T13:51:03.428-04:00 | result = conn._execute_20(statement, params or {}, execution_options)
  | 2021-09-16T13:51:03.428-04:00 | File "/pymodules/sqlalchemy/engine/base.py", line 1583, in _execute_20
  | 2021-09-16T13:51:03.428-04:00 | return meth(self, args_10style, kwargs_10style, execution_options)
  | 2021-09-16T13:51:03.428-04:00 | File "/pymodules/sqlalchemy/sql/elements.py", line 323, in _execute_on_connection
  | 2021-09-16T13:51:03.428-04:00 | return connection._execute_clauseelement(
  | 2021-09-16T13:51:03.428-04:00 | File "/pymodules/sqlalchemy/engine/base.py", line 1452, in _execute_clauseelement
  | 2021-09-16T13:51:03.428-04:00 | ret = self._execute_context(
  | 2021-09-16T13:51:03.428-04:00 | File "/pymodules/sqlalchemy/engine/base.py", line 1814, in _execute_context
  | 2021-09-16T13:51:03.428-04:00 | self._handle_dbapi_exception(
  | 2021-09-16T13:51:03.428-04:00 | File "/pymodules/sqlalchemy/engine/base.py", line 1995, in _handle_dbapi_exception
  | 2021-09-16T13:51:03.428-04:00 | util.raise_(
  | 2021-09-16T13:51:03.428-04:00 | File "/pymodules/sqlalchemy/util/compat.py", line 207, in raise_
  | 2021-09-16T13:51:03.428-04:00 | raise exception
  | 2021-09-16T13:51:03.428-04:00 | File "/pymodules/sqlalchemy/engine/base.py", line 1771, in _execute_context
  | 2021-09-16T13:51:03.428-04:00 | self.dialect.do_execute(
  | 2021-09-16T13:51:03.428-04:00 | File "/pymodules/sqlalchemy/engine/default.py", line 717, in do_execute
  | 2021-09-16T13:51:03.428-04:00 | cursor.execute(statement, parameters)
  | 2021-09-16T13:51:03.428-04:00 | sqlalchemy.exc.OperationalError: (psycopg2.OperationalError) SSL connection has been closed unexpectedly
  | 2021-09-16T13:51:03.428-04:00 | [SQL: SELECT lists.id AS lists_id
  | 2021-09-16T13:51:03.428-04:00 | FROM lists
  | 2021-09-16T13:51:03.428-04:00 | WHERE lists.service_id = %(service_id_1)s]
```

--- a/api/database/db.py
+++ b/api/database/db.py
@@ -12,5 +12,8 @@ db_engine = create_engine(
     connection_string,
     pool_size=1,
     max_overflow=0,
+    pool_pre_ping=True,  # Check that a connection is still active before attempting to use
+    pool_recycle=1500,  # Prune connections older than 25 minutes (RDS Proxy has a timeout of 30 minutes)
+    pool_use_lifo=True,  # Always re-use last connection used (allows server-side timeouts to remove unused connections)
 )
 db_session = sessionmaker(bind=db_engine)

